pkg/controller/operators/olm/operator.go

```diff
@@ -102,12 +103,57 @@ type Operator struct {
 	clientFactory                clients.Factory
 	plugins                      []plugins.OperatorPlugin
 	informersByNamespace         map[string]*plugins.Informers
+	informersFiltered            bool
+
+	ruleChecker     func(*v1alpha1.ClusterServiceVersion) *install.CSVRuleChecker
+	ruleCheckerLock sync.RWMutex
+	resyncPeriod    func() time.Duration
+	ctx             context.Context
 }
 
 func (a *Operator) Informers() map[string]*plugins.Informers {
 	return a.informersByNamespace
 }
 
+func (a *Operator) getRuleChecker() func(*v1alpha1.ClusterServiceVersion) *install.CSVRuleChecker {
+	var ruleChecker func(*v1alpha1.ClusterServiceVersion) *install.CSVRuleChecker
+	a.ruleCheckerLock.RLock()
+	ruleChecker = a.ruleChecker
+	a.ruleCheckerLock.RUnlock()
+	if ruleChecker != nil {
+		return ruleChecker
+	}
+
+	a.ruleCheckerLock.Lock()
+	defer a.ruleCheckerLock.Unlock()
+	if a.ruleChecker != nil {
+		return a.ruleChecker
+	}
+
+	sif := informers.NewSharedInformerFactoryWithOptions(a.opClient.KubernetesInterface(), a.resyncPeriod())
+	rolesLister := sif.Rbac().V1().Roles().Lister()
+	roleBindingsLister := sif.Rbac().V1().RoleBindings().Lister()
+	clusterRolesLister := sif.Rbac().V1().ClusterRoles().Lister()
+	clusterRoleBindingsLister := sif.Rbac().V1().ClusterRoleBindings().Lister()
+
+	done := make(chan struct{})
+	go func() {
+		<-a.ctx.Done()
+		done <- struct{}{}
+	}()
+	sif.Start(done)
+	sif.WaitForCacheSync(done)
```

Comment:
It returns a map of sync status, but this function blocks until all caches are synced, or the stop channel has been closed.

Reply:
Right, so there's no difference in what we do between the two possible outcomes?
- caches synced successfully
- the stop channel was closed

Reply:
Typically you're waiting on the factory to sync before parking your main goroutine via `<-ctx.Done()`. But because here we aren't in the main goroutine and this function is returning a value, I agree we should act differently if the stop channel is closed. What does that look like, though - return nil and handle the nil upstream, panic, ...?

Reply:
Just crash the process & restart? Done context is cancelled when we get SIGINT or similar

Reply:
I probably wouldn't `os.Exit` here in case there's some cleanup code (now or in the future) that thinks it's going to get a chance to run.

Reply:
What should we do other than return from the function?

Reply:
This is only used by `func (a *Operator) permissionStatus(strategyDetailsDeployment *v1alpha1.StrategyDetailsDeployment, targetNamespace string, csv *v1alpha1.ClusterServiceVersion) (bool, []v1alpha1.RequirementStatus, error) {`, I think, so we could have that function check the return value, and if it's nil, return an error. Or this function could also return an error.